docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,319 @@
+   Now, `DependenciesForArtifact` serves two purposes. As before, the map keys are relative local paths of input files for the container.
+   In addition, the map values list all absolute remote destination paths where the input files are placed in the container.
+   This has to be a list, because a single input file may be copied to several locations in the container.
+   
+   Thus, `DependenciesForArtifact` serves two purposes: firstly, it lists all input files to watch for changes.
+   Secondly, it provides default sync destinations for changed or deleted files.
+   Note that this does not mean, that every file will be sync'ed. If a file is entitled for syncing depends on the user-provided sync rules.
+   
+   For example, given the above Dockerfile, it will return something like
+   ```go
+   map[string][]string{
+    "static/page1.html": []string{"/var/www/page1.html"},
+    "static/assets/style.css": []string{"/var/www/assets/style.css"},
+    "index.html": []string{"/var/www/index.html"},
+    "nginx.conf": []string{"/var/nginx.conf", "/etc/nginx.conf"},
+   }
+   ```
+   
+2. To support additional sync maps from builders, the builder interface will need an additional method `SyncMap`:
+   ```go
+   type Builder interface {
+       Labels() map[string]string
+
+       Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+
+       DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) (map[string][]string, error)
+    
+       // ADDED
+       SyncMap() map[string]string
+   }
+   ```
+   
+   The new `SyncMap` function shall return pairs of
+   
+   - **key**: a glob pattern of local files to sync into the container
+   - **value**: a destination path in the container
+   
+   By convention, the sync rule from a builder does never flatten directories but does subdir syncing instead (also see open question below).
+
+### Config changes
+
+The `artifact.sync` config needs to support additional requirements:
+
+- Support existing use-cases for backwards compatibility, i.e.
+  - Glob patterns on the host path
+  - Transplant a folder structure, or flatten the folder structure
+- Offer a way to use destination inference.
+- Avoid ambiguity
+
+#### Problems with existing syntax
+- The flattening of the file structure at the destination is surprising. For example:
+  ```yaml
+  sync:
+    'src/**/*.js': dest/
+  ```
+  will put all `.js` files in the `dest` folder.
+- The triple-star syntax is is quite implicit and a surprising extension of the standard globbing syntax. For example:
+  ```yaml
+  sync:
+    'src/b/c/***/*.py': dest/
+  ```
+  will recreate subfolders _below_ `src/b/c` at `dest/` and sync `.py` files.
+  The triple-star therefore triggers two actions which should not depend on each other:
+  - do not flatten the directory structure
+  - strip everything on the source path up to `***`
+
+#### Suggestion
```

Comment:
My thoughts about this alternative:
- Multiple destinations should be a rare exception. An exception we must to handle internally, but none that we need to expose to our users. If somebody wants to sync to multiple destinations, she can as well add another rule for the same src pattern.
- Should the user have to enable magic sync manually? I think this would be easily forgotten, because most users would not be aware of this. IMHO magic sync should be implicitly on, if available (and not overridden by a manual sync spec). So I would not consider this for the pipeline config.
- Finally, I'm not a big fan of the explicit `inferDest`. But maybe it makes the feature more discoverable.
@briandealwis @tejal29 WDYT?

Reply:
@corneliusweig and @balopat i am little bit skeptical about the key `direct`. Its confusing to me, maybe `manual` makes it better.

> Multiple destinations should be a rare exception.

Agree with this. If a file syncing to multiple directories is a side-effect of multiple rules, it ok.

@balopat  with your suggestion, would this  skaffold config be valid? 
```
sync: 
  manual: 
     - src: 'special-directory/**/*.js'
       strip: "special-directory/"  # default "" 
       dest: "dir1/"
  infer:
      - 'src/**/*.js'
      - 'src/**/*.html'
```
I really don't have a preference to be honest. This is something we can find out with Critical User Journeys.
But yes, you do end up writing less config with suggested change. Maybe that a win?

Reply:
> * Should the user have to enable magic sync manually? I think this would be easily forgotten, because most users would not be aware of this. IMHO magic sync should be implicitly on, if available (and not overridden by a manual sync spec). So I would not consider this for the pipeline config.

Filesync is a serious business - we are mutating the user's container. I think sync should be opt-in, and even if we get comfortable with it and turn it on by default after a while, I believe there will be users who want to be able to turn it off. 

I'm okay with `manual` maybe instead of `explicit`, I do like `infer`. 

Multiple dest: I'm convinced to not exposed that to the user, sounds good to keep dest single valued.

Reply:
discoverability: totally true - in IntelliJ, VSCode this structure will help a lot for users to find the options.